### PR TITLE
PLFM-6641: Worker to process unlinked file handles

### PR DESCRIPTION
**Pull request by marcomarasca**
Depends on https://github.com/Sage-Bionetworks/Synapse-Stack-Builder/pull/382

### CODE REVIEW

**Comment by marcomarasca** [comment]
Depends on https://github.com/Sage-Bionetworks/Synapse-Stack-Builder/pull/382

File: services/workers/src/main/resources/athena-spb.xml [hunk 1]
```diff
@@ -44,4 +44,40 @@
...
+			<bean class="org.springframework.scheduling.quartz.MethodInvokingJobDetailFactoryBean">
+				<property name="targetObject">
+					<bean class="org.sagebionetworks.workers.util.aws.message.MessageDrivenWorkerStack">
+						<constructor-arg index="0" ref="countingSemaphore" />
+						<constructor-arg index="1" ref="awsSQSClient" />
+						<constructor-arg index="2">
+							<bean class="org.sagebionetworks.workers.util.aws.message.MessageDrivenWorkerStackConfiguration">
+								<property name="runner" ref="recurrentAthenaQueryWorker" />
+								<property name="semaphoreLockKey" value="recurrentAthenaQueryWorker" />
+								<property name="semaphoreMaxLockCount" value="5" />
```
Comment (john-hill) [inline services/workers/src/main/resources/athena-spb.xml:64]: maybe only one for now

File: services/workers/src/main/resources/athena-spb.xml [hunk 2]
```diff
@@ -44,4 +44,40 @@
 		<property name="repeatInterval" value="86400000" />
 	</bean>
 	
+	<!-- This worker process an SQS message that contains the query results from an Athena query executed by a step function -->
+	<bean id="recurrentAthenaQueryWorker" class="org.sagebionetworks.athena.workers.RecurrentAthenaQueryWorker" scope="singleton" >
+		<property name="queueName" value="#{stackConfiguration.getQueueName('RECURRENT_ATHENA_QUERIES')}" />
```
Comment (john-hill) [inline services/workers/src/main/resources/athena-spb.xml:49]: can pull out a reusable property for the queue name.

File: services/workers/src/main/java/org/sagebionetworks/file/worker/FileHandleStreamWorker.java [hunk 3]
```diff
@@ -59,7 +58,7 @@ public void run(ProgressCallback progressCallback, List<ChangeMessage> messages)
 			return;
 		}
 		
-		kinesisLogger.logBatch(STREAM_NAME, records);
+		kinesisLogger.logBatch(FileHandleRecord.STREAM_NAME, records);
 		
 	}
```
Comment (john-hill) [inline services/workers/src/main/java/org/sagebionetworks/file/worker/FileHandleStreamWorker.java:64]: either add the real status of the file handle or only export the available.

File: services/workers/src/test/java/org/sagebionetworks/file/worker/FileHandleUnlinkedQueryIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+			return new Pair<>(linked.size() == 1 && unlinked.size() == 1, null);
+		});
+	}
+	
+	private void waitForKinesisData(List<String> ids, String tableName, String idColumn) throws Exception {
+		Database dataBase = athenaSupport.getDatabase("firehoseLogs");
+		
+		Table fileHandleDataTable = athenaSupport.getTable(dataBase, tableName);
+		
+		String query = "SELECT COUNT(*) FROM " + fileHandleDataTable.getName() + " WHERE " + idColumn + " IN (" + String.join(",", ids) + ")";
```
Comment (john-hill) [inline services/workers/src/test/java/org/sagebionetworks/file/worker/FileHandleUnlinkedQueryIntegrationTest.java:197]: we have asynch test helpers that could have this logic:
retun athenaQuery(query, callbac)

File: services/workers/src/test/java/org/sagebionetworks/file/worker/FileHandleUnlinkedQueryIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+			}
+			
+			LOG.info("Executing query {}...DONE (Count: {})", query, count);
+			
+			return new Pair<>(count >= ids.size(), null);
+		});
+	}
+	
+	private String findStateMachineArn(String name) {
+		ListStateMachinesResult result;
```
Comment (john-hill) [inline services/workers/src/test/java/org/sagebionetworks/file/worker/FileHandleUnlinkedQueryIntegrationTest.java:224]: might want to make the get reusable.

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/DBOFileHandleDaoImplTest.java [hunk 6]
```diff
@@ -829,4 +833,64 @@ public void testIsMatchingMD5WithBothNull() {
...
+		// Only the second is available now
+		assertEquals(Arrays.asList(Long.valueOf(file1.getId())), result);
+	}
+	
+	@Test
+	public void testUpdateBatchStatus() {
+		
+		Date createdOn = Date.from(Instant.now().minusSeconds(60));
+		
+		S3FileHandle file1 = TestUtils.createS3FileHandle(creatorUserGroupId, idGenerator.generateNewId(IdType.FILE_IDS).toString()).setCreatedOn(createdOn);
```
Comment (john-hill) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/DBOFileHandleDaoImplTest.java:870]: we have a DBOHelpler<FileHandle>

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/athena/RecurrentAthenaQueryManager.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+import org.sagebionetworks.repo.model.athena.RecurrentAthenaQueryResult;
+
+import com.amazonaws.services.sqs.model.Message;
+
+public interface RecurrentAthenaQueryManager {
+
+	RecurrentAthenaQueryResult fromSqsMessage(Message message);
+
+	void processRecurrentAthenaQueryResult(RecurrentAthenaQueryResult request, String queueUrl);
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/athena/RecurrentAthenaQueryManager.java:11]: comments?

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/athena/RecurrentAthenaQueryProcessor.java [hunk 8]
```diff
@@ -0,0 +1,15 @@
+package org.sagebionetworks.repo.manager.athena;
+
+import java.util.List;
+
+import org.sagebionetworks.repo.model.athena.RowMapper;
+
+public interface RecurrentAthenaQueryProcessor<T> {
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/athena/RecurrentAthenaQueryProcessor.java:7]: comments

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/athena/RecurrentAthenaQueryManagerImpl.java [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+		if (!AthenaQueryExecutionState.SUCCEEDED.equals(execution.getState())) {
+			throw new IllegalStateException("The query with id " + queryExecutionId + " did not SUCCEED (State: " + execution.getState() + ", reason: " + execution.getStateChangeReason() + ")");
+		}
+		
+		int pageRequests = 0;
+		
+		// The initial page token is null since it's coming from the step function that does not include any (e.g. first page)
+		String currentPageToken = request.getPageToken();
+
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/athena/RecurrentAthenaQueryManagerImpl.java:110]: comment would help why are we not finishing?

**Review by john-hill — APPROVED** [review]

**Pull request merged**